src/configs/nodeclassification_generators/sbm/default_setup.gin

```diff
@@ -15,3 +15,11 @@ SbmGeneratorWrapper.param_sampler_specs = %SBMSamplerSpec
 
 # Normalize node features (default true)
 SbmGeneratorWrapper.normalize_features = True
+
+# Use the communities generated by the LFR model (default false)
+SbmGeneratorWrapper.use_generated_lfr_communities = False
+
+# Define parameters for LFR model if use_generated_lfr_communities set to True
+SbmGeneratorWrapper.lfr_params = {'avg_degree':15, 'max_degree_proportion': 0.17, 'power_exponent':-3, 
```

Comment:
This looks better! However, to make it even cleaner, please find a way to pass in the `LFRSamplerSpec` here instead of manually inputting the values. Import src/configs/nodeclassification_generators/lfr/default_setup.gin at the top of this file. Then make `lfr_params` in the SbmGeneratorWrapper accept the `LFRSamplerSpec` list instead.

Reply:
Something else to consider here: it seems like `community_min_size_proportion`,  `community_max_size_proportion`, and `community_exponent` are three parameters that a GraphWorld user should be able to modify when defining an SBM experiment with LFR communities. Is there a way to make these configurable parameters for SBM (to be used *only* when LFR communities are imputed), and make the rest of the parameters auxiliary (pulled from a default LFR config)?

Reply:
@jpalowitch For a GraphWorld user to modify those 3 parameters on an SBM experiment with LFR communities, do you mean that they should be able to define them in the config for SBM - as opposed to these particular parameters being pulled from the LFR config.

So a user can vary just these 3 parameters within SBM's config, whilst using different values for the same parameters in LFR's config for an LFR experiment?

Reply:
> This looks better! However, to make it even cleaner, please find a way to pass in the `LFRSamplerSpec` here instead of manually inputting the values. Import src/configs/nodeclassification_generators/lfr/default_setup.gin at the top of this file. Then make `lfr_params` in the SbmGeneratorWrapper accept the `LFRSamplerSpec` list instead.

Hey @jpalowitch thanks for the pointers! I'm able to call the LFR params from the LFR config by using: 

`include '....\lfr\default_setup.gin'` and `SbmGeneratorWrapper.lfr_params = %LFRSamplerSpec`

which works for fine for SBM with LFR communities. However, it also causes an override of the shared parameters between LFR and SBM anytime the include statement is executed. So even if running a regular SBM experiment it will still use the shared params from LFR's config like nvertex and avg degree. 

Thus it requires commenting out those 2 lines in the default setup for SBM when wanting to run a regular experiment. This seems to be the only way to [import a .gin into another .gin file](https://github.com/google/gin-config/blob/master/docs/index.md#including-other-gin-files), but let me know if I'm missing something.

One solution would be to prefix all the LFR parameters with `'lfr_'` so there would be no overlapping parameters that are overriden, and this could potentially be useful in analysing the data of SBM with LFR communities to know what parameters came from where?

Wondering if you have any idea for a cleaner solution? Or if the commenting out is sufficient?

Reply:
Hey Mustafa, very thoughtful questions, thank you for laying out some of the pros and cons! The first thing to note here is that prefixing the _gin scopes_ of the generator parameters will not affect their _parameter names_ in the output GraphWorld dataframe. In a gin config line such as `nvertex/ParamSamplerSpec.name = "nvertex"`, the first `nvertex` is the gin scope, and the second `nvertex` is what the parameter is named.

So, we should definitely prefix the LFR parameter gin scopes with `lfr_` in this case, since we now want to import that config, and it causes gin naming conflicts. We should _not_ however change the variable names, since for any given parameter in the output dataframe, it should always be clear how it was used. Any LFR-specific parameter is only used when the `SbmGeneratorWrapper.use_generated_lfr_communities` gin parameter is set to `True`. In summary:

* Prefix the LFR parameter definitions with `lfr_`. I would suggest _not_ doing this for all other generators -- for now let's only do this for files that cause import conflicts.
* Make sure all SBM parameters and all SBM-relevant LFR parameters end up in the output dataframe, even when `SbmGeneratorWrapper.use_generated_lfr_communities` is set to `False`, so that there is consistency across experiments.

Reply:
Hey @jpalowitch. My solution in the latest commit is to add all relevant (non-overlapping) LFR parameters to the CABAM/SBM parameter specs, in the [Generator Wrappers init ](https://github.com/google-research/graphworld/blob/d1726611dd33e0b165972c1221e53506bc5b269c/src/graph_world/nodeclassification/generator_wrappers.py#L31)regardless of if using LFR communities or not. 

[The function that performs this](https://github.com/google-research/graphworld/blob/d1726611dd33e0b165972c1221e53506bc5b269c/src/graph_world/beam/generator_config_sampler.py#L99) will also override the overlapping parameter values between LFR and CABAM/SBM so that LFR's parameters are prioritised. However, I can simplify this so that LFR's parameter ranges are only used to _supplement_, the existing CABAM/SBM parameters rather than override them, when generating graphs of this type.

These new parameters added to the spec will then only be sampled if using LFR communities. So that when not using LFR communities [they have null values in the output.](https://github.com/google-research/graphworld/blob/d1726611dd33e0b165972c1221e53506bc5b269c/src/graph_world/beam/generator_config_sampler.py#L93) I did this as I thought it would be confusing to have sampled values for LFR in the output for CABAM/SBM when these values were not used in the graph generation.

Reply:
Hi Mustafa, thanks for the latest commits. This is even closer, but still not quite what GraphWorld should be doing in this case. I have made some specific low-level code suggestions to get us there. At a high-level, this is what we want:

(1) If `use_generated_lfr_communities` is false, the SBM pipeline should run exactly as it did before this PR -- no extra parameter samplers added to the class internally, no extra parameter values outputted in the dataframe externally
(2) If `use_generated_lfr_communities` is true, the 4 LFR parameters that matter to community generation should be added to the SBM generator wrapper for sampling (see my suggestions on src/graph_world/nodeclassification/generator_wrappers.py for this). All other LFR parameters should adopt their default values in the LFR default_param_ranges gin file to generate the LFR graph.

Reply:
All of this should also be done for the CABAM pipeline, in the same way.